compiler/src/dotty/tools/dotc/core/GadtConstraint.scala

```diff
@@ -209,7 +209,15 @@ final class ProperGadtConstraint private(
   def isNarrowing: Boolean = wasConstrained
 
   override def approximation(sym: Symbol, fromBelow: Boolean)(using Context): Type = {
-    val res = approximation(tvarOrError(sym).origin, fromBelow = fromBelow)
+    val res =
+      approximation(tvarOrError(sym).origin, fromBelow = fromBelow) match
+        case tpr: TypeParamRef =>
+          // Here we do externalization when the returned type is a TypeParamRef,
+          //  b/c ConstraintHandling.approximation may return internal types when
+          //  the type variable is instantiated. See #15531.
+          externalize(tpr)
+        case tp => tp
```

Comment:
Refactor idea: why don't we pull `externalizeMap` out of `nonParamBounds`, maybe even do the usual lazy instantiation trick (search for "theAcc" or "theMap") so we can process all types not just `TypeParamRef`s?

Reply:
👍  Yes, it would be good to pull out `externalizeMap` (and apply the lazy instantiation trick) and enable `externalize` to handle all kinds of types. 

But in terms of the issue to be fixed here, externalizing other types in `approximation` may not be necessary, since `ConstraintHandling` will call `fullLowerBound` in other cases, which is overridden in `GadtConstraint` to do externalization properly. So the externalization is already done unless the returned type is a `TypeParamRef`.

Shall we do the refactorization in a subsequent PR? Or shall we just do it in this one?

Reply:
I'm happy to do a follow up PR if you prefer.  So I can also look at those details you mentioned and we don't have to think about them right now. 😄